google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/StreamingSubscriberConnection.java

```diff
@@ -211,35 +211,45 @@ private boolean isAlive() {
   @Override
   public void sendAckOperations(
       List<String> acksToSend, List<PendingModifyAckDeadline> ackDeadlineExtensions) {
+    StreamingPullRequest.Builder[] requests =
+        partitionAckOperations(acksToSend, ackDeadlineExtensions, MAX_PER_REQUEST_CHANGES);
+    for (StreamingPullRequest.Builder request : requests) {
+      requestObserver.onNext(request.build());
+    }
+  }
 
-    // Send the modify ack deadlines in batches as not to exceed the max request
-    // size.
-    List<List<String>> ackChunks = Lists.partition(acksToSend, MAX_PER_REQUEST_CHANGES);
-    List<List<PendingModifyAckDeadline>> modifyAckDeadlineChunks =
-        Lists.partition(ackDeadlineExtensions, MAX_PER_REQUEST_CHANGES);
-    Iterator<List<String>> ackChunksIt = ackChunks.iterator();
-    Iterator<List<PendingModifyAckDeadline>> modifyAckDeadlineChunksIt =
-        modifyAckDeadlineChunks.iterator();
-
-    while (ackChunksIt.hasNext() || modifyAckDeadlineChunksIt.hasNext()) {
-      StreamingPullRequest.Builder requestBuilder =
-          StreamingPullRequest.newBuilder();
-      if (modifyAckDeadlineChunksIt.hasNext()) {
-        List<PendingModifyAckDeadline> modAckChunk = modifyAckDeadlineChunksIt.next();
-        for (PendingModifyAckDeadline modifyAckDeadline : modAckChunk) {
-          for (String ackId : modifyAckDeadline.ackIds) {
-            requestBuilder
-                .addModifyDeadlineSeconds(modifyAckDeadline.deadlineExtensionSeconds)
-                .addModifyDeadlineAckIds(ackId);
-          }
-        }
-      }
-      if (ackChunksIt.hasNext()) {
-        List<String> ackChunk = ackChunksIt.next();
-        requestBuilder.addAllAckIds(ackChunk);
+  @VisibleForTesting
+  static StreamingPullRequest.Builder[] partitionAckOperations(
+      List<String> acksToSend, List<PendingModifyAckDeadline> ackDeadlineExtensions, int size) {
+    int numExtensions = 0;
+    for (PendingModifyAckDeadline modify : ackDeadlineExtensions) {
+      numExtensions += modify.ackIds.size();
+    }
+    int numRequests = Math.max(numExtensions, acksToSend.size());
+    numRequests = numRequests / size + (numRequests % size == 0 ? 0 : 1);
+
+    StreamingPullRequest.Builder[] requests = new StreamingPullRequest.Builder[numRequests];
+    for (int i = 0; i < numRequests; i++) {
+      requests[i] = StreamingPullRequest.newBuilder();
+    }
+
+    int count = 0;
+    for (List<String> acksChunk : Lists.partition(acksToSend, size)) {
+      requests[count].addAllAckIds(acksChunk);
+      count++;
+    }
+
+    count = 0;
+    for (PendingModifyAckDeadline modify : ackDeadlineExtensions) {
+      for (String ackId : modify.ackIds) {
+        requests[count / size]
```

Comment:
what is `count / size` supposed to compute? If `size` is `10000` (`MAX_PER_REQUEST_CHANGES`), I have trouble seeing when this will ever be anything besides 0. 

Now I really want to see a test for this...